Enforce Linting/Prettier

amirhmk opened the pull request:
Add `ESLinting` and configurations to the project. Have setup a ESLint which follows "`Airbnb`'s rules" for design and best practises. 

Have also most fixed most errors that to reduce the clutter, though there is still about 2-3 different errors which we can have a look at. I tried figuring out the `imports extensions` rule, even disabling the rule did not make it go away. Help on that would be great. 

Another question is we are currently including TensorFlow directly from the HTML as a `<script>`, are we not including the library itself because of performance reasons? (Not sure if that applies even)

Did not make any modularization changes, but I think we can look into that as well.

I also tried making ESLint work for the HTML file, my setup is a little buggy, can someone confirm this works for that as well?

**Important**:

Have added a new nom command, `lint`. We should try to run this every time before a commit to make sure our code is nice and clean. Should probably be added as a part of CI too if we want to be rigour with clean code.

Berkmann18 commented:
I haven't dealt with AirBnB's import extension in the past but looking at https://github.com/airbnb/javascript, I can see that it should be fairly easy to fix (since it's just about file names and using more concise imports).

I also noticed the codebase has redundant imports which aren't tree-shaken which would lead to worse performance and heavier script load in the browser.
So I think, we should definitely look into the modularity of the whole codebase so modules only import (and export) what is needed and we could also consider using a bundler.

Berkmann18 commented on js/main.js:
```diff
@@ -39,62 +51,65 @@ async function initCamera() {
+  initVideoUI();
+  const canvas = initCanvas();
 
   startEngine(canvas, video);
```
Not really related to this PR, but ESLint complains that video isn't defined ("'video' is not defined  no-undef") so I think we could fix that in this PR alongside other ESLint errors. What do you think?

I looked at the code and I believe it was meant to be `state.video` instead of `video` (@misterpeddy correct me if I'm wrong).

amirhmk replied:
Great, thank you!!

amirhmk commented:
> I haven't dealt with AirBnB's import extension in the past but looking at https://github.com/airbnb/javascript it should be easy, however, removing the `.js` in the imports lead to errors on the browser (so perhaps bundling the code would help?).
> 
> I also noticed the codebase has redundant imports which aren't tree-shaken which would lead to worse performance and heavier script load in the browser.
> So I think, we should definitely look into the modularity of the whole codebase so modules only import (and export) what is needed and we could also consider using a bundler (be it webpack, Rollup, Parcel, Browserify, ...).
> 
> As for TF.js, I think using the CDN approach would be better performance-wise but I could be wrong.
> 
> ESLint is by nature only meant for JS/TS/JSON. So for HTML and CSS we should use other tools such as Tidy and Stylelint respectively.

Thank you for the comments. Very helpful actually 💯 

So yeah I was thinking about bundling and `webpack` too, it's a good change. I think we should do it, Webpack could work? also do you think we should have it in this PR or address it in its own PR? happy to look into that either way👍 

Tidy I will look into further, came across it but wasn't sure, thanks!!

Berkmann18 commented:
> So yeah I was thinking about bundling and webpack too, it's a good change. I think we should do it, Webpack could work?

Yeah, although we could also consider Parcel or Rollup (both of which scored high in satisfaction and popularity behind Webpack, on [StateOfJS 2019](https://2019.stateofjs.com/other-tools/#build_tools) and [BestOfJS](https://bestofjs.org/projects?tags=module)).

I think it would be better to have in another PR since it's not closely related to code style stuff that linters/formatters handle and I guess some may prefer using other bundlers.

amirhmk commented:
@Berkmann18 @misterpeddy I have tidy so it can validate HTML files and fix them as much as it can. Tidy should be installed on Macs I think, but it's not just like a `devDependency` thats you would get after installing the modules. It's more like you globally install it once on the machine (if needed), and then use it. [Tidy](http://www.html-tidy.org) is the link to getting it if you don't already have it

Anyways, I have added a config file that was provided by the official doc, and added the command to fix it in the existing `list` npm command. Please have a look, if all good I'll go ahead with the merge.

Parcel looks fast and robust in our case, let's try them out for the bundling up the module task!

Berkmann18 commented:
There's actually a [NPM package](https://www.npmjs.com/package/tidy) for Tidy but it's old and possibly not maintained.
In any case, I found the package Arch Repo and installed it and I'm sure other distros/OSs have it.

After digging in old projects, I just realised that Prettier **can** format HTML pages (and seems to do a good job, I forgot that it could do that).

And if we were to use tools that could be installed via NPM alone (so easier for people working on the project _and_ the CI/commit-hook pipeline which hopefully will have a linting/formatting step), we could use [`html-validate`](https://www.npmjs.com/package/html-validate) (which is configurable and in the same spirit as ESLint).

misterpeddy commented on js/main.js:
```diff
@@ -1,13 +1,25 @@
-import * as draw from './draw.js';
```
Interesting - I'm assuming this was the linter? 
since the functions are no longer referred to from their module's namespace, can we rename frame to drawFrame?

amirhmk replied:
yep makes sense, will do!

misterpeddy commented:
> @Berkmann18 @misterpeddy I have tidy so it can validate HTML files and fix them as much as it can. Tidy should be installed on Macs I think, but it's not just like a `devDependency` thats you would get after installing the modules. It's more like you globally install it once on the machine (if needed), and then use it. [Tidy](http://www.html-tidy.org) is the link to getting it if you don't already have it
> 
> Anyways, I have added a config file that was provided by the official doc, and added the command to fix it in the existing `list` npm command. Please have a look, if all good I'll go ahead with the merge.
> 
> Parcel looks fast and robust in our case, let's try them out for the bundling up the module task!

just got it! for debian, it's distributed through apt so you're right that we just have to rely on people getting it. Do people usually put this in the readme/development section or is it basically assumed that people already have it on their machine?

amirhmk commented:
Okay dope guys, I think I have addressed all the comments, `stylelint` is now added and an additional check on lint. 

As per `markdown` lint, I wasn't so sure if we need it as a package on the project, I have updated the `README` to reflect that the user should make sure to do this, like I'm saying this because users may have different linters - I use the VSCode one. Let me know if you have any other concerns

There is a `nom warning` which is addressed [here](https://www.npmjs.com/advisories/1179) because of `minimalist`. It's a new issue and should be a small patch, will keep an eye on that as well

Berkmann18 commented:
Nice catch on the vulnerability (which still need to be patched in `stylelint` dependencies.
I've started working on the PR with the fix (although it will need to wait on the relevant releases).
